### PR TITLE
Popup menu button alignment fix

### LINKED ISSUE #26360: Offset of PopupMenuButton takes selected item into account
I noticed my `PopUpMenu` jumping around whenever I selected any value in it. After deeper inspection I noticed the `PopUpMenu`'s offset is linked to the `initialValue`. But this doesn't make any sense and is also not documented.

I've build a small example app to show it. The `CheckedPopUpMenu` acts normally to its given offset. But the other `PopUpMenu`'s offset is only correct when the first item is selected. If any other item is selected it jumps to the top off the screen (because the offset of the selected item is offscreen probably).
The checkbox in the middle adds the workaround I currently use, which is adding more offset relative to the selected item.

Example app code to be copy-pasted into e.g. Dartpad:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'PupUpMenu bug',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'PupUpMenu bug'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({required this.title, super.key});

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final _itemCount = 4;
  final _checkedItems = [false, false, false, false];

  int? _selectedIndex;
  bool _applyWorkaround = false;

  @override
  Widget build(BuildContext context) {
    final _checkedItemWidgets = List.generate(_itemCount, (index) {
      return CheckedPopupMenuItem(
        child: Text("Check $index"),
        checked: _checkedItems[index],
        value: index,
      );
    });

    final _itemWidgets = List.generate(_itemCount, (index) {
      return PopupMenuItem(
        child: Text("Check $index"),
        value: index,
      );
    });

    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
        actions: [
          PopupMenuButton(
            icon: Icon(Icons.check),
            offset: Offset(0, 100),
            onSelected: (index) {
              setState(() {
                _checkedItems[index] = !_checkedItems[index];
              });
            },
            itemBuilder: (_) => _checkedItemWidgets,
          ),
          PopupMenuButton(
            initialValue:
                _selectedIndex, // Setting this makes the offset focus on this item
            icon: Icon(Icons.close),
            offset: Offset(
                0,
                100 +
                    (_applyWorkaround ? 44 * (_selectedIndex ?? 0) : 0)
                        .toDouble()),
            onSelected: (index) {
              setState(() {
                _selectedIndex = index;
              });
            },
            itemBuilder: (_) => _itemWidgets,
          )
        ],
      ),
      body: Center(
        child: Row(
          mainAxisSize: MainAxisSize.max,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Checkbox(
                value: _applyWorkaround,
                onChanged: (checked) {
                  if (checked != null) {
                    setState(() {
                      _applyWorkaround = checked;
                    });
                  }
                }),
            SizedBox(width: 8.0),
            Text("Apply workaround"),
          ],
        ),
      ),
    );
  }
}
```

Flutter doctor output:

```
Running flutter doctor...
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel master, v1.1.9-pre.25, on Microsoft Windows [Version 10.0.17134.472], locale nl-NL)
[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[√] Android Studio (version 3.2)
[√] Android Studio (version 3.3)
[√] Connected device (1 available)

• No issues found!

```

### LINKED ISSUE #115509: PopupMenuButton isn't aligned correctly when initialValue isn't null.
The alignment of the PopupMenuButton changes based on whether the InitialValue attribute is set to a non-null value.

When initialValue is null, the position of the overlay is exactly on top of the popupMenuButton. When initialValue is set to a value, the overlay is above the popupMenuButton + so extra amount.

<details>
<summary>Video before the fix:</summary>

https://user-images.githubusercontent.com/73116038/202327937-32b36186-3ce7-4169-9fb8-654c4c5f31f3.mp4

</details>

<details>
<summary>Video after the fix</summary>

https://user-images.githubusercontent.com/73116038/202328078-1677e12b-1fa1-4a94-93b2-a679a9103a79.mp4

</details>

<details>

<summary>Reproducible code:</summary>

```

import 'package:flutter/material.dart';

// This is the type used by the popup menu below.
enum Menu { itemOne, itemTwo, itemThree, itemFour }

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: _title,
      home: MyStatefulWidget(),
    );
  }
}

class MyStatefulWidget extends StatefulWidget {
  const MyStatefulWidget({super.key});

  @override
  State<MyStatefulWidget> createState() => _MyStatefulWidgetState();
}

class _MyStatefulWidgetState extends State<MyStatefulWidget> {
  String _selectedMenu = '';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
      ),
      body: Column(
        children: [
          Row(
            children: [
              PopupMenuButton<Menu>(
                // Callback that sets the selected popup menu item.
                  onSelected: (Menu item) {
                    setState(() {
                      _selectedMenu = item.name;
                    });
                  },
                  itemBuilder: (BuildContext context) => <PopupMenuEntry<Menu>>[
                    const PopupMenuItem<Menu>(
                      value: Menu.itemOne,
                      child: Text('Item 1'),
                    ),
                    const PopupMenuItem<Menu>(
                      value: Menu.itemTwo,
                      child: Text('Item 2'),
                    ),
                    const PopupMenuItem<Menu>(
                      value: Menu.itemThree,
                      child: Text('Item 3'),
                    ),
                    const PopupMenuItem<Menu>(
                      value: Menu.itemFour,
                      child: Text('Item 4'),
                    ),
                  ]),

              SizedBox(width: 100),

              PopupMenuButton<Menu>(
                // Callback that sets the selected popup menu item.
                initialValue: Menu.itemOne,
                  onSelected: (Menu item) {
                    setState(() {
                      _selectedMenu = item.name;
                    });
                  },
                  itemBuilder: (BuildContext context) => <PopupMenuEntry<Menu>>[
                    const PopupMenuItem<Menu>(
                      value: Menu.itemOne,
                      child: Text('Item 1'),
                    ),
                    const PopupMenuItem<Menu>(
                      value: Menu.itemTwo,
                      child: Text('Item 2'),
                    ),
                    const PopupMenuItem<Menu>(
                      value: Menu.itemThree,
                      child: Text('Item 3'),
                    ),
                    const PopupMenuItem<Menu>(
                      value: Menu.itemFour,
                      child: Text('Item 4'),
                    ),
                  ]),
            ],
          ),
          Text('_selectedMenu: $_selectedMenu'),
        ],
      ),
    );
  }
}

```
</details>

I am on the master channel

### PR DESCRIPTION
**Pull request by whiskeyPeak**
I modified the original code to include an additional fix for the popupMenuItems moving based on which index is selected. This change makes it so that the popupMenuItems always come out right below the popupMenuButton.

https://user-images.githubusercontent.com/73116038/204927662-37b43193-3fe9-4261-9f8d-b846b0b30de6.mp4

fixes #115509 
fixes #26360

### CODE REVIEW

File: packages/flutter/lib/src/material/popup_menu.dart
Comment (QuncCccccc) [inline packages/flutter/lib/src/material/popup_menu.dart:700]: With this change, would the menu still show up below the button even if the `position` is set to `PopupMenuPosition.over`?

**Review by QuncCccccc — COMMENTED** [review]
Thanks a lot for your contribution! Just left a question below. Please make sure that there are some tests for the code change and all existing tests should pass.

File: packages/flutter/lib/src/material/popup_menu.dart
Comment (whiskeyPeak) [reply packages/flutter/lib/src/material/popup_menu.dart]: > With this change, would the menu still show up below the button even if the position is set to PopupMenuPosition.over?

No, not unless an Offset is provided.

If the position attribute is set to PopupMenuPosition.over, the menu will be drawn above it, if it is set to PopupMenuPosition.under, it will be drawn under it.

https://user-images.githubusercontent.com/73116038/205382980-217b0b25-2299-4535-9149-1ad1f4ee4761.mp4

**Comment by whiskeyPeak** [comment]
@QuncCccccc Added a test, PTAL!

File: packages/flutter/lib/src/material/popup_menu.dart
Comment (QuncCccccc) [inline packages/flutter/lib/src/material/popup_menu.dart:698]: nit: probably we can remove this comment as we don't have to calculate the vertical position.

**Review by QuncCccccc — COMMENTED** [review]
Looks good! Thanks a lot. Just left one more comment below.

File: packages/flutter/lib/src/material/popup_menu.dart
Comment (whiskeyPeak) [reply packages/flutter/lib/src/material/popup_menu.dart]: Done!

**Review by QuncCccccc — APPROVED** [review]
LGTM. Thanks!

**Comment by whiskeyPeak** [comment]
@QuncCccccc Hey, does this pr need another reviewer or can the auto-submit label be added to it?

**Review by justinmc — COMMENTED** [review]
I want to double check that the code that was removed is indeed no longer needed.  It looks like what it's trying to do is to align the selected item vertically with the anchor.  Are we sure that's not desired behavior?

The [Material spec](https://m3.material.io/components/menus/overview) shows a bit about these menus but I don't think it definitively answers the question.  Actually in this screenshot it looks like the position of the menu is below the anchor (with no selection), which doesn't seem to be what Flutter is doing anyway:

screenshot of Material website showing popup menu

@whiskeyPeak I see that you were [asking about this](https://github.com/flutter/flutter/issues/26360#issuecomment-1332152572) and the correct behavior but never got an answer.  Maybe we need to come up with a full plan on how this widget should behave before we change anything.

*Edit:* These are the problems that I see with PopupMenu positioning.  We should probably make decisions about how all of these should work before merging any changes.

  1. The default position (with nothing selected) probably should be below the anchor and not on top of it, based on the Material spec.
  2. We should decide if the selected item move to align itself with the anchor.  If so, then the current implementation probably isn't right, based on https://github.com/flutter/flutter/issues/26360.  When the PopupMenuButtons are given an offset, it might be incorrect.
  3. There should be a reasonable API for adjusting the position of this menu.  The way that this is done in https://github.com/flutter/flutter/issues/26360 seems hacky and seems to break things.

**Review by QuncCccccc — CHANGES_REQUESTED** [review]
Oh, I agree with what @justinmc said. Just removing all this block of code might not be a good idea.

**Comment by whiskeyPeak** [comment]
@justinmc Thanks for the review!

> The default position (with nothing selected) probably should be below the anchor and not on top of it, based on the Material spec.

I agree. I had a look at the material 2 and 3 spec and that seems to be the default behaviour.

> We should decide if the selected item move to align itself with the anchor. If so, then the current implementation probably isn't right, based on https://github.com/flutter/flutter/issues/26360. When the PopupMenuButtons are given an offset, it might be incorrect.

I had a look at the available spec again, and there doesn't seem to be an indication that it should move based on what item is currently selected. Also, given that this type of behaviour isn't really expected and that an issue has been filled about it, I think it makes more sense to have the offset stay constant.
There is a flutter widget called DropDownButton which uses this changed offset, though in this context it make sense since the button becomes part of the overlay menu

https://user-images.githubusercontent.com/73116038/206317083-34a3f890-3532-48d7-a8eb-a5f4f0079734.mp4


> There should be a reasonable API for adjusting the position of this menu. The way that this is done in https://github.com/flutter/flutter/issues/26360 seems hacky and seems to break things.

There already is. You can use the Offset parameter to re-adjust the x and y position of the menu

File: packages/flutter/test/material/popup_menu_test.dart
Comment (whiskeyPeak) [reply packages/flutter/test/material/popup_menu_test.dart]: I updated the tests to what I think you meant? I'm somewhat new to this test business haha

**Comment by bleroux** [comment]
@whiskeyPeak To fix the failing tests you probably have to rebase your branch on master (instead of using merge) and forced push.

**Comment by justinmc** [comment]
@whiskeyPeak Are you still available to rebase this branch and get it passing?

**Pull request closed without merging**
